pull: Refactor pull_layers()

jodh-intel opened the pull request:
Split the `pull_layers()` method which now calls `handle_layer()` to
reduce the length and complexity of the original function.

Also implemented `TryFrom` for `Compression` (and added some unit tests)
to further simplify the code.

> **Note:**
>
> This PR also contains a new `assert_result!()` test macro that was
> originally written by me as `assert_log_level!()` for Kata Containers
> on:
>
> - https://github.com/kata-containers/kata-containers/pull/65
>
> ... but which now exists in Kata Containers as `assert_result!()` here:
>
> - https://github.com/kata-containers/kata-containers/blob/2b5dc2ad39b9b077246e5bc17c91cc2578c22497/src/libs/test-utils/src/lib.rs#L68

Fixes: #54.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

jodh-intel commented:
Note that this PR conflicts with #52. If that one lands first, I'll rebase to add the fix on that PR to the new `handle_layer()` method on this PR...

jodh-intel commented:
ftr, I think the pull code could / should be simplfied further. This is just the start ;)

arronwy commented on src/pull.rs:
```diff
@@ -83,108 +85,138 @@ impl PullClient {
+    }
+}
+
+impl TryFrom<&str> for Compression {
```
Can we move the implementation and tests of `TryFrom` for `Compression` to https://github.com/confidential-containers/image-rs/blob/main/src/decoder/mod.rs ?

jodh-intel replied:
Good call - done.

arronwy reviewed (commented):
Thanks @jodh-intel LGTM, just one suggestion inline.

arronwy commented:
> Note that this PR conflicts with #52. If that one lands first, I'll rebase to add the fix on that PR to the new `handle_layer()` method on this PR...

Yes, #52 is just merged, please help rebase with that fix.

jodh-intel commented:
@arronwy - I've updated the PR by:

- Creating a local crate for the `assert_result!()` macro so it can be used easily from different parts of the code.
- Added some tests for `handle_layer()`.

Note the single cached `PullClient()` in `test_handle_layer()` to avoid continually hitting the network.

arronwy reviewed (approved):
Thanks @jodh-intel , LGTM!

jodh-intel commented:
Resolved conflicted manifest file.
